lib/Serializer.ts

```diff
@@ -0,0 +1,120 @@
+import { ObjectType, ModelType } from "./General";
+import { Document } from "./Document";
+
+const defaultSerializer = {
+	"modify": (serialized: ObjectType, original: ObjectType): ObjectType => {
+		return {...original};
+	}
+};
+
+const includeHandler = (document: ModelType<Document>, includeRules: string[], serialized: ObjectType): ObjectType => {
+	return includeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(document).includes(key)) {
+			serialized[key] = document[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const excludeHandler = (document: ModelType<Document>, excludeRules: string[], serialized: ObjectType): ObjectType => {
+	return excludeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(serialized).includes(key)) {
+			delete serialized[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const validateName = (name: string): void => {
+	if (!name || typeof name !== "string") {
+		throw new Error("Field name is required and should be of type string");
+	}
+};
+
+const validateOptions = (options): void => {
+	if (!options || !(Array.isArray(options) || typeof options === "object")) {
+		throw new Error("Field options is required and should be an object or array");
+	}
+};
+
+const cleanAndValidateDocumentsArray = (documentsArray: ModelType<Document>[]): ModelType<Document>[] => {
+	if (!documentsArray || !Array.isArray(documentsArray)) {
+		throw new Error("documentsArray must be an array of document objects");
+	}
+
+	return documentsArray.filter((doc) => typeof doc.serialize === "function");
+};
+
+export class Serializer {
+	#serializers: any;
+	#defaultSerializer: string;
+
+	constructor() {
+		this.#serializers = {_default: defaultSerializer};
+		this.#defaultSerializer = "_default";
+	}
+
+	add(name: string, options): void {
```

Comment:
We have to add a TypeScript interface file for options. I have to go back and reread the original proposal and dig into what is allowed to be passed into `options`.

Reply:
Fixed in 76c139d6808cefc95a7d6b66ef1981a405936301